web/app.js

```diff
@@ -0,0 +1,80 @@
+var express = require("express");
+var path = require("path");
+var logger = require("morgan");
+var cookieParser = require("cookie-parser");
+var bodyParser = require("body-parser");
+var { uniq, compact, sortBy } = require("lodash");
+var app = express();
+
+// uncomment after placing your favicon in /public
+//app.use(favicon(__dirname + "/public/favicon.ico"));
+app.use(logger("dev"));
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({ extended: false }));
+app.use(cookieParser());
+app.use(express.static(path.join(__dirname, "public")));
+
+var server = require("http").Server(app);
+var io = require("socket.io")(server);
+
+var cats = require("../");
+
+io.on("connection", function (socket) {
+
+  socket.emit("line", "Accessing San Francisco SPCA (Cat Department)...");
+
+  cats.fetchCats()
+    .then(uniq)
+    .map(url => {
+      return cats.urlToCat(url)
+        .tap(cat => {
+          socket.emit("line", `Weighing cat: ${cat.name}`);
+        });
+    }, { concurrency: 10 })
+    .then(compact)
+    .then(cats => {
+      const fattestCat = sortBy(cats, "weight")[cats.length - 1];
+      socket.emit("line", `The fattest cat is ${fattestCat.name}. ${(fattestCat.isFemale ? "She" : "He")} weighs ${fattestCat.lbs} lbs and ${fattestCat.oz} oz.`);
```

Comment:
You could have the whole pipeline as as single function which takes a side-effect-producer as it's argument. Then the caller is responsible for chaining on to the end of the function to do redirect logic.